baseapp/baseapp_test.go

```diff
@@ -98,6 +98,54 @@ func setupBaseApp(t *testing.T, options ...AppOption) *BaseApp {
 	return app
 }
 
+// baseapp loaded from a fraudproof
+func setupBaseAppFromFraudProof(t *testing.T, fraudProof FraudProof, options ...AppOption) *BaseApp {
+	storeKeys := make([]stypes.StoreKey, 0, len(fraudProof.stateWitness))
+	routerOpts := make([]func(bapp *BaseApp), 0, len(fraudProof.stateWitness))
+	for storeKeyName := range fraudProof.stateWitness {
+		storeKey := sdk.NewKVStoreKey(storeKeyName)
+		storeKeys = append(storeKeys, storeKey)
+
+		routerOpt := func(bapp *BaseApp) {
+			bapp.Router().AddRoute(sdk.NewRoute(routeMsgKeyValue, func(ctx sdk.Context, msg sdk.Msg) (*sdk.Result, error) {
+				kv := msg.(*msgKeyValue)
+				bapp.cms.GetKVStore(storeKey).Set(kv.Key, kv.Value)
+				return &sdk.Result{}, nil
+			}))
+		}
+		routerOpts = append(routerOpts, routerOpt)
+	}
+	options = append(options, SetSubstores(storeKeys...))
+	// RouterOpts should only be called after call to `SetSubstores`
+	for _, routerOpt := range routerOpts {
+		options = append(options, AppOptionFunc(routerOpt))
+	}
+
+	// This initial height is used in `BeginBlock` in `validateHeight`
+	options = append(options, SetInitialHeight(fraudProof.blockHeight))
+
+	for _, storeKey := range storeKeys {
```

Comment:
Why do we need a second loop?  Can everything be done in the first loop by getting the `witness` in addition to the `storeKeyName`?

Reply:
Great catch, I initially thought the ordering was relevant but on a second look, `SetSubstoreKVPair` is an `AppOptionOrdered` so it's only called after the `SetSubstores` have been initialized.